draft-rescorla-tls-esni.md

```diff
@@ -601,8 +601,52 @@ lie. In this design, the attacker would need to be able to find a
 Z which would expand into a key that would validly AEAD-encrypt
 a message of his choice, which should be intractable (Hand-waving alert!).
 
+# Alternative SNI Protection Designs
 
-# Alternate Encryption Design
+Alternative approaches to encrypted SNI may be implemented at the TLS or 
+application layer. In this section we describe several alternatives and discuss 
+drawbacks in comparison to the design in this document.
+
+## TLS-layer
+
+### TLS Tunnels
+
+In this variant, real TLS Client Hellos are tunneled within the early data payload 
+of an outer TLS connection established with the client-facing server. This requires 
+the client to have established a previous session — and obtained a PSK — with the 
```

Comment:
I am not sure if "TLS tunnel" requires a previous session or the use of 0-RTT data.

Would you mind considering either adjusting the name or adjusting the text? Because what I expect for a scheme called "TLS tunnel" is as follows:

"TLS tunnel" is an approach that first establishes a TLS connection to a fronting host, and then, on top of that connection, establishes another TLS connection that uses the hostname of the hidden host.

If that is the case, I'd assume that all prior knowledge that is necessary is the binding between two names: name of the fronting host and that of the hidden host.

In other words, I'd expect that the "TLS Tunnel" scheme will be distributing using DNS the name of the fronting host instead of the ESNI key.

The downside of the tunneling is approach will be that there will be additional roundtrip when 0-RTT is not used, as well as the cost of double handshake and double encryption.

Reply:
The tunnel approach you describe is a simple proxy. We can describe it here, though it's different from what was the "hide TLS in early data" proposal. I updated the title to try to make it more clear.